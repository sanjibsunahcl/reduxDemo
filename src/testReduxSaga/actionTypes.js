export const userConstants = {
  LOAD_USERS: 'LOAD_USERS',
  LOAD_USERS_FAIL: 'LOAD_USERS_FAIL',
  LOAD_PROFILES: 'LOAD_PROFILES',
  LOAD_PROFILES_FAIL: 'LOAD_PROFILES_FAIL',
  GET_USERS: 'GET_USERS',
  GET_USER_PROFILES: 'GET_USER_PROFILES',
};
