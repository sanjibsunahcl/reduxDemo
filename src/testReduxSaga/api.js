export const api = {
  users: {
    async getUsers(params) {},
    async getProfiles(params) {},
  },
};

export function errorHandler(err) {}
