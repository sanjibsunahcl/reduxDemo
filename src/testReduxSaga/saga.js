import {call, put, takeLatest} from 'redux-saga/effects';
import {api, errorHandler} from './api';
import {userConstants} from './actionTypes';

export function* getUsers(action) {
  try {
    const {id} = action.payload;
    const {data} = yield call(api.users.getUsers, {id});
    console.log('user' + JSON.stringify(data));
    yield put({type: userConstants.LOAD_USERS, payload: data});
  } catch (err) {
    yield call(errorHandler, err);
    yield put({type: userConstants.LOAD_USERS_FAIL});
  }
}

export function* getProfiles(action) {
  try {
    const {id} = action.payload;
    const {data} = yield call(api.users.getProfiles, {id});
    console.log('useProfile' + JSON.stringify(data));
    yield put({type: userConstants.LOAD_PROFILES, payload: data});
  } catch (err) {
    yield call(errorHandler, err);
    yield put({type: userConstants.LOAD_PROFILES_FAIL});
  }
}

export function* usersSaga() {
  yield takeLatest(userConstants.GET_USERS, getUsers);
  yield takeLatest(userConstants.GET_USER_PROFILES, getProfiles);
}
