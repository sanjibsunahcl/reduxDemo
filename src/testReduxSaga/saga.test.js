import {testSaga, expectSaga} from 'redux-saga-test-plan';
import {getProfiles, getUsers, usersSaga} from './saga';
import {userConstants} from './actionTypes';
import {api, errorHandler} from './api';

describe('test Saga users', () => {
  describe('usersSaga', () => {
    it('should pass', () => {
      testSaga(usersSaga)
        .next()
        .takeLatest(userConstants.GET_USERS, getUsers)
        .next()
        .takeLatest(userConstants.GET_USER_PROFILES, getProfiles)
        .finish()
        .isDone();
    });
  });

  describe('getUsers', () => {
    it('should get data', () => {
      testSaga(getUsers, {payload: {id: 1}})
        .next()
        .call(api.users.getUsers, {id: 1})
        .next({data: 'teresa teng'})
        .put({type: userConstants.LOAD_USERS, payload: 'teresa teng'})
        .finish()
        .isDone();
    });
    it('should handle error', () => {
      const err = new Error('network');
      testSaga(getUsers, {payload: {id: 1}})
        .next()
        .call(api.users.getUsers, {id: 1})
        .throw(err)
        .call(errorHandler, err)
        .next()
        .put({type: userConstants.LOAD_USERS_FAIL})
        .finish()
        .isDone();
    });
  });

  describe('getProfiles', () => {
    it('should get profile', () => {
      testSaga(getProfiles, {payload: {id: 1}})
        .next()
        .call(api.users.getProfiles, {id: 1})
        .next({data: 'teresa teng'})
        .put({type: userConstants.LOAD_PROFILES, payload: 'teresa teng'})
        .finish()
        .isDone();
    });
    it('should handle error', () => {
      const err = new Error('network');
      testSaga(getProfiles, {payload: {id: 1}})
        .next()
        .call(api.users.getProfiles, {id: 1})
        .throw(err)
        .call(errorHandler, err)
        .next()
        .put({type: userConstants.LOAD_PROFILES_FAIL})
        .finish()
        .isDone();
    });
  });

  describe('Ui testing', () => {
    it('snapshot testing', () => {
      return expectSaga(getProfiles, {payload: {id: 1}})
        .run()
        .then(result => {
          expect(result.toJSON()).toMatchSnapshot();
        });
    });
  });
});
