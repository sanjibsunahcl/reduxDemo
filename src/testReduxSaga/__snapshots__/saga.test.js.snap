// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Saga users Ui testing snapshot testing 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          Object {
            "id": 1,
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/getProfiles",
      },
      "type": "CALL",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          Object {},
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/errorHandler",
      },
      "type": "CALL",
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "type": "LOAD_PROFILES_FAIL",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
  ],
}
`;
